Add lib/ansible/db_simple.py

sfromm opened the pull request:
db_simple.py is a basic db abstraction for storing results
from ansible runs via a callback.  The database has one table,
host_result, and the columns: timestamp, hostname, module (name of
module that was run, if available), result (success, failure, etc), and
data (the json result data).  While handling the data column, data
is either json encoded or decoded via json.dumps or json.loads.

I'll follow with callback_plugins/sql.py.  This is related to #2008.  Please comment on schema and I can adjust it as necessary.  :-)

sfromm committed:
Add lib/ansible/db_simple.py

db_simple.py is a basic db abstraction for storing results
from ansible runs via a callback.  The database has one table,
host_result, and the columns: timestamp, hostname, module (name of
module that was run, if available), result (success, failure, etc), and
data (the json result data).  While handling the data column, data
is either json encoded or decoded via json.dumps or json.loads.

skvidal commented:
A couple of comments from discussion on irc:
- tracking the playbook being run would be helpful
  - full path to the playbook
  - optional git commit checksum if playbook is in a git repo
  - optional md5sum of playbook if the above is not true
- tracking the user running the playbook (os.getlogin() and os.geteuid())
- tracking the playbook (and play) being run per action logged into the above db.
- tracking the complete set of options passed in on either cli or via api for the playbook
  - extra_vars
  - config file state (connection)
- tracking the execution timing per-action and/or per-play
- tracking if the action was called using an alternative connection type (fireball)

With this information we could do a lot with processing and analyzing playbook executions as well as producing some great log/report outputs :)

sfromm committed:
Update db_simple with user info

Add AnsibleUser that stores the username and effective user of the
person running ansible.  This also renames AnsibleResult to AnsibleTask.

sfromm commented:
Added the user information to db_simple.py.  I think we could add `start` and `end` to what module_common.py reports in exit_json, similar to what the `command` module does now.  The task table could then be augmented to include that if available.

The as-yet-unpublished sql.py callback creates AnsibleUser, if it doesn't already exist, and associates the task with that user.

@skvidal, I think many of the playbook tracking pieces you propose are good ideas.  Just need that functionality added there.

mpdehaan commented:
As mentioned on IRC, I suspect the callback API will soon track a hell of a
lot more data soon, making a lot more be possible.

When we do acom (and we're doing acom), we'll evaluate that.

We obviously have to keep the old callback API around for compatibility and
have to extend it in smart ways.

On Thu, Feb 7, 2013 at 6:43 PM, Stephen Fromm notifications@github.comwrote:

> Added the user information to db_simple.py. I think we could add startand
> end to what module_common.py reports in exit_json, similar to what the
> command module does now. The task table could then be augmented to
> include that if available.
> 
> The as-yet-unpublished sql.py callback creates AnsibleUser, if it doesn't
> already exist, and associates the task with that user.
> 
> @skvidal https://github.com/skvidal, I think many of the playbook
> tracking pieces you propose are good ideas. Just need that functionality
> added there.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2009#issuecomment-13269245.

mpdehaan commented:
As I also mentioned, I will probably be rather picky on what the schema
ends up being, but there is ZERO reason we can't have this also as a
callback ansible.

However, that all being said, this cannot live in lib/ansible right now, we
shouldn't have any database awareness in core ansible.

As an example plugin, yes.

On Thu, Feb 7, 2013 at 8:54 PM, Michael DeHaan michael.dehaan@gmail.comwrote:

> As mentioned on IRC, I suspect the callback API will soon track a hell of
> a lot more data soon, making a lot more be possible.
> 
> When we do acom (and we're doing acom), we'll evaluate that.
> 
> We obviously have to keep the old callback API around for compatibility
> and have to extend it in smart ways.
> 
> On Thu, Feb 7, 2013 at 6:43 PM, Stephen Fromm notifications@github.comwrote:
> 
> > Added the user information to db_simple.py. I think we could add startand
> > end to what module_common.py reports in exit_json, similar to what the
> > command module does now. The task table could then be augmented to
> > include that if available.
> > 
> > The as-yet-unpublished sql.py callback creates AnsibleUser, if it doesn't
> > already exist, and associates the task with that user.
> > 
> > @skvidal https://github.com/skvidal, I think many of the playbook
> > tracking pieces you propose are good ideas. Just need that functionality
> > added there.
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2009#issuecomment-13269245.

mpdehaan commented:
closing for above reasons, let's do this as an example callback -- feel free to continue discussion, I still think this is great and am NOT trying to squash this effort.

Just be aware I may not 100% reference it later, but it should be able to continue to work.

sfromm commented:
While I understand lib/ansible may not be the perfect spot, it really helps
to have the schema separate so that both callbacks and scripts to later
parse the db can import the same file.  It just so happens that lib/ansible
is awfully convenient.

Speaking extemporaneously, would it be possible to have a db_plugins in
core and then let users drop in schemas/whatever that work for them?

On Thu, Feb 7, 2013 at 5:55 PM, Michael DeHaan notifications@github.comwrote:

> As I also mentioned, I will probably be rather picky on what the schema
> ends up being, but there is ZERO reason we can't have this also as a
> callback ansible.
> 
> However, that all being said, this cannot live in lib/ansible right now,
> we
> shouldn't have any database awareness in core ansible.
> 
> As an example plugin, yes.
> 
> On Thu, Feb 7, 2013 at 8:54 PM, Michael DeHaan michael.dehaan@gmail.comwrote:
> 
> > As mentioned on IRC, I suspect the callback API will soon track a hell
> > of
> > a lot more data soon, making a lot more be possible.
> > 
> > When we do acom (and we're doing acom), we'll evaluate that.
> > 
> > We obviously have to keep the old callback API around for compatibility
> > and have to extend it in smart ways.
> > 
> > On Thu, Feb 7, 2013 at 6:43 PM, Stephen Fromm notifications@github.comwrote:
> > 
> > > Added the user information to db_simple.py. I think we could add
> > > startand
> > > end to what module_common.py reports in exit_json, similar to what the
> > > command module does now. The task table could then be augmented to
> > > include that if available.
> > > 
> > > The as-yet-unpublished sql.py callback creates AnsibleUser, if it
> > > doesn't
> > > already exist, and associates the task with that user.
> > > 
> > > @skvidal https://github.com/skvidal, I think many of the playbook
> > > tracking pieces you propose are good ideas. Just need that
> > > functionality
> > > added there.
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub<
> > > https://github.com/ansible/ansible/pull/2009#issuecomment-13269245>.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2009#issuecomment-13273343.

## 

Stephen Fromm

mpdehaan commented:
Unfortunately I don't really want schemas to be pluggable, and if we are to
have a schema in core, I'm going to be very picky about it.   Ansible
doesn't need a database nor know about them, things that use ansible are
allowed to (and those include plugins).

You can of course write a logging plugin that imports any python modules
you want, it probably just needs to live outside core then -- that could be
any module it does not need to even be in the ansible namespace.

When we're resurrecting acom itself, acom _itself_ will live outside of
core (including plugins for it).

I'd suggest if this is to be an example of a DB callback I would put the
model in the same file.

On Thu, Feb 7, 2013 at 11:12 PM, Stephen Fromm notifications@github.comwrote:

> While I understand lib/ansible may not be the perfect spot, it really
> helps
> to have the schema separate so that both callbacks and scripts to later
> parse the db can import the same file. It just so happens that lib/ansible
> is awfully convenient.
> 
> Speaking extemporaneously, would it be possible to have a db_plugins in
> core and then let users drop in schemas/whatever that work for them?
> 
> On Thu, Feb 7, 2013 at 5:55 PM, Michael DeHaan notifications@github.comwrote:
> 
> > As I also mentioned, I will probably be rather picky on what the schema
> > ends up being, but there is ZERO reason we can't have this also as a
> > callback ansible.
> > 
> > However, that all being said, this cannot live in lib/ansible right now,
> > we
> > shouldn't have any database awareness in core ansible.
> > 
> > As an example plugin, yes.
> > 
> > On Thu, Feb 7, 2013 at 8:54 PM, Michael DeHaan michael.dehaan@gmail.comwrote:
> > 
> > > As mentioned on IRC, I suspect the callback API will soon track a hell
> > > of
> > > a lot more data soon, making a lot more be possible.
> > > 
> > > When we do acom (and we're doing acom), we'll evaluate that.
> > > 
> > > We obviously have to keep the old callback API around for
> > > compatibility
> > > and have to extend it in smart ways.
> > > 
> > > On Thu, Feb 7, 2013 at 6:43 PM, Stephen Fromm <
> > > notifications@github.com>wrote:
> > > 
> > > > Added the user information to db_simple.py. I think we could add
> > > > startand
> > > > end to what module_common.py reports in exit_json, similar to what
> > > > the
> > > > command module does now. The task table could then be augmented to
> > > > include that if available.
> > > > 
> > > > The as-yet-unpublished sql.py callback creates AnsibleUser, if it
> > > > doesn't
> > > > already exist, and associates the task with that user.
> > > > 
> > > > @skvidal https://github.com/skvidal, I think many of the playbook
> > > > tracking pieces you propose are good ideas. Just need that
> > > > functionality
> > > > added there.
> > > > 
> > > > —
> > > > Reply to this email directly or view it on GitHub<
> > > > https://github.com/ansible/ansible/pull/2009#issuecomment-13269245>.
> > 
> > —
> > Reply to this email directly or view it on GitHub<
> > https://github.com/ansible/ansible/pull/2009#issuecomment-13273343>.
> 
> ## 
> 
> Stephen Fromm
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/2009#issuecomment-13276246.
